### PR TITLE
CSHARP-3840: Unresponsive/deadlocked cluster.Dispose()

### PR DESCRIPTION
**Pull request by sanych-sun**

### CODE REVIEW

File: src/MongoDB.Driver/Core/Connections/TcpStreamFactory.cs [hunk 1]
```diff
@@ -122,103 +122,76 @@ private void ConfigureConnectedSocket(Socket socket)
...
-                        await Task.Factory.FromAsync(socket.BeginConnect(dnsEndPoint.Host, dnsEndPoint.Port, null, null), socket.EndConnect).ConfigureAwait(false);
-                    }
-                    else
-                    {
-                        await Task.Factory.FromAsync(socket.BeginConnect(endPoint, null, null), socket.EndConnect).ConfigureAwait(false);
-                    }
+            if (endPoint is DnsEndPoint dnsEndPoint)
+            {
+                // mono doesn't support DnsEndPoint in its BeginConnect method.
+                connectTask = Task.Factory.FromAsync(socket.BeginConnect(dnsEndPoint.Host, dnsEndPoint.Port, null, null), socket.EndConnect);
```
Comment (BorisDog) [inline src/MongoDB.Driver/Core/Connections/TcpStreamFactory.cs:169]: Please check if `ConnectAsync` can replace `BeginConnect` (`SocketTaskExtensions`)

File: src/MongoDB.Driver/Core/Connections/TcpStreamFactory.cs [hunk 2]
```diff
@@ -122,103 +122,76 @@ private void ConfigureConnectedSocket(Socket socket)
 
         private void Connect(Socket socket, EndPoint endPoint, CancellationToken cancellationToken)
         {
-            var state = 1; // 1 == connecting, 2 == connected, 3 == timedout, 4 == cancelled
+            IAsyncResult connectOperation;
```
Comment (BorisDog) [inline src/MongoDB.Driver/Core/Connections/TcpStreamFactory.cs:125]: Please evaluate the option of using `ConnectAsync()....GetResult()` instead.

**Review by BorisDog — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver/Core/Connections/TcpStreamFactory.cs [hunk 3]
```diff
@@ -122,103 +122,62 @@ private void ConfigureConnectedSocket(Socket socket)
...
-                {
-                    var message = string.Format("Timed out connecting to {0}. Timeout was {1}.", endPoint, _settings.ConnectTimeout);
-                    throw new TimeoutException(message);
-                }
-                if (state == 4) { throw new OperationCanceledException(); }
+            WaitHandle.WaitAny([connectOperation.AsyncWaitHandle, cancellationToken.WaitHandle], _settings.ConnectTimeout);
+
+            if (!connectOperation.IsCompleted)
+            {
+                socket.Dispose();
```
Comment (BorisDog) [inline src/MongoDB.Driver/Core/Connections/TcpStreamFactory.cs:141]: I think we need to hide `Dispose` exceptions here, and probably in all cases?

File: src/MongoDB.Driver/Core/Connections/TcpStreamFactory.cs [hunk 4]
```diff
@@ -122,103 +122,62 @@ private void ConfigureConnectedSocket(Socket socket)
...
-                catch when (state == 1)
-                {
-                    try { socket.Dispose(); } catch { }
-                    throw;
-                }
-                catch when (state >= 3)
-                {
-                    // a timeout or operation cancelled exception will be thrown instead
-                }
+                socket.Dispose();
```
Comment (BorisDog) [inline src/MongoDB.Driver/Core/Connections/TcpStreamFactory.cs:167]: Ignore exceptions?

**Review by BorisDog — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver/Core/Connections/TcpStreamFactory.cs (hunk 3 above)
Comment (sanych-sun) [reply src/MongoDB.Driver/Core/Connections/TcpStreamFactory.cs]: Done

File: src/MongoDB.Driver/Core/Connections/TcpStreamFactory.cs (hunk 4 above)
Comment (sanych-sun) [reply src/MongoDB.Driver/Core/Connections/TcpStreamFactory.cs]: Done

**Review by BorisDog — APPROVED** [review]
LGTM!

**Pull request merged**